### DISCUSSION TITLE
Integrating Humidity Switch US-SS02 into home automation

### DISCUSSION THREAD

**Discussion by sun2sirius** [arendst/Tasmota#21009] [Q&A] [answered]
I installed this Humidity Switch to control a bathroom fan in a rental unit. I have smart devices of other brands that integrate into my SmartThings system at this location. I also have a bunch of devices with Tasmota firmware in my main home where I have a SmartThings hub and all works great. But I did not want to install another SmartThings hub in the rental place and instead use the US-SS02 device's MQTT or other remote control capabilities. I already spent hours searching and trying - this should not be that hard. As I understand, the Home Assistance integration requires a local server - that is not an option. I tried finding a cloud-based MQTT broker - the only one I was able to connect to was MyQttHub - OK, I get the periodic messages from the sensor. But there seems to be no way to organize those messages and make sense/use of them unless I write a whole client using their API and pull the messages out of the queue. 
 
Are there really no existing systems that do this? All I really need is to get the sensor readings, create and update rules remotely, and control the switch if necessary. Can someone please suggest a few options how I can remotely access that switch? I do not mind to write some code and experiment integrating with different systems, but would like to get some pointers how to do it right.

**Comment by sfromis** [comment]
If you have MQTT access, that's the way you can create/update rules using commands, and subscribe to the sensor data automatically published. Any command you can issue locally in Tasmota can also be issued via MQTT, including `Power on` or `Power off`. 

And if you configure the Tasmota device appropriately, including an unique topic, it should all automatically be organized neatly per device. Of course, if you can manage it via local Tasmota rules, that's fine, and easy to update via MQTT.

**Reply by sun2sirius** [reply]
Thank you so much, I keep on MQTT right now, but it does not go easy - just need a bit more help. First, although I do see messages from the device in MyQttHub, I still cannot find a way to pull them from MyQttHub. I know, it would be the MyQttHub support question, but maybe MyQttHub is not the best choice to work with Tasmota, maybe you can recommend a working use case with a walkthrough or something? Second, suppose I can subscribe for MQTT messages - I assume I would need a pointer to the right place in Tasmota documentation, where it would describe how to update the rules via MQTT as you are suggesting. Thanks again.

**Reply by sfromis** [reply]
I can't talk about how MyQttHub specifically works, but assuming that it "simply" acts as a MQTT broker on the Internet, the usual way would be that both devices (or many devices) connect to the same broker, and then each end can publish and subscribe by topic.

And things like updating rules is exactly the same as you can do in the console, except that you publish the commands via MQTT. Tasmota will automatically subscribe to the command topic, and execute incoming commands.
https://tasmota.github.io/docs/MQTT/#commands-over-mqtt

Of course, when it gets on the big bad public internet, you should be using the TLS feature for encrypted communication.

I'd recommend first getting the hang of it by having a local broker (like Mosquitto), before going outside the home. Of course, as with any concept, there is a learning curve to get started, and the page I linked above does go into how Tasmota uses MQTT.

**Reply by barbudor** [reply]
A couple of years ago I tested various online MQTT services including MyQttHub and a couple of others before choosing Flespi.io
I don't remember having a hard time with MyQTTHub and I was for sure using MQTT over TLS.
Not saying anything bad about MyQTTHHub, just that it was checking less boxes than Flespi.io for my use-case.

I don't know what a SmartThings hub needs as I'm using NodeRED but I have no difficulties using my main NodeRED at main house with both my local mosquitto and the cloud-based MQTT.
The same way you should be able to have an HA instance in your main home but talking to the cloud-MQTT server. Just don't choose the mode where HA use it's own broker but use a separated broker. I have no idea how to use that as I'm not a HA user.

To analyse MQTT communications, tools like MQTT Explorer or MQTT-X will work the exact same way with a cloud based MQTT.

Still don't forget that if you control things in your remote rental biulding though the cloud from a HA at home, you are dependant on the good operation of 2 different Internet connections. If one fails, you are loosing all control.

**Reply by sfromis** [reply]
Beware that having "smart" devices in a rental unit means that the tenant has physical access, and hence opportunities for messing around with the device, possibly doing things you'd not like. And the tenant might have "strong negative" views about the landlord having devices on the Wifi network there, with concerns like privacy.

**Reply by sun2sirius** [reply]
I tried Flespi.io too, but for some reason the device would not connect. First it gives error 5, then starts retrying and reports error -2. By now as I received your guys' responses, I was able to pull messages from MyQttHub, was even able to send the cmnd/tasmota_switch/Power message to the device through the broker and get the response. It looks like I will need to write my own client to talk to MyQttHub as they have a bit different authentication than the ready clients I tried have - you first login with username and password to get a token, then use the token for all other commands, then logout and destroy the token.

I fly back home in a couple of days and just need to be sure that the device will keep the MQTT connection to the MyQttHub broker, so that I can get back to it from home when I can finish the client side. Will the device be persistent at calling the broker, e.g. if power goes out, etc. anomalies? And another Q - is it possible to update the MQTT configuration itself over MQTT, like the Host and username/password? Is there any failsafe mechanism to go back to the previous config or something?

@sfromis - yes you are so correct about the rental place's environment and privacy/security issues associated with that :) I hope to sort it out after I have the technical part figured out. Thank you all for the great info, I can report back if I fully get this to work.

**Reply by barbudor** [reply]
As I said, I can connect any of my tools to Flespi so I don't see why you would be able to do the same with MyMqttHub.
If you can connect a Tasmota device to MyMqttHub, there's no reason you couldn't connect a HA or Nodered. 


If the settings are good, tasmota will always reconnects properly after a power outage


There is no built-in failsafe in Tasmota to revert to a 'known good' mqtt configuration if you loose access after a remote change of configuration.
It should be possible to do one by using rules such as:
- starting a rule timer on wifi#connected
- on mqtt#connected cancel the timer (set to 0)
- on wifi#disconnected cancel also the timer
- on timer trigger, issue a backlog command with known good configuration

Once the rule is in place (better test it before leaving) you can remotely change the MQTT configuration. If mqtt doesn't connect it will revert the settings and reboot

Using wifi#connected is better than system#boot because if wifi is down, you don't want to trigger the revert

**Reply by barbudor** [reply]
> It looks like I will need to write my own client to talk to MyQttHub as they have a bit different authentication than the ready clients I tried have - you first login with username and password to get a token, then use the token for all other commands, then logout and destroy the token.

I don't understand this part
I just created a device for MqttExplorer
- Device Id: MqttExp0001
- Username: MqttExplorer
- Password: *******

Then I setup the same in MqttExplorer and I was able to connect simply
No need for temporary token

The 1st device is the admin account
Then I have 1 tasmota
And finally the account for MqttExplorer
I could do the same for HA, NodeRED, .... as far as you stick in the limits of your plan

![image](https://github.com/arendst/Tasmota/assets/2996042/6581a5f5-3937-435b-aa82-d6cf41c6a7d2)

**Comment by barbudor** [comment]
Hi @sun2sirius 
Have you been successful ?

**Comment by sun2sirius** [comment]
Yes, I was about to share my results but you beat me to it :) 
I was able to connect that US-SS02 humidity switch with Tasmota firmware to MyQttHub and regularly collect its humidity reports. I am also able to turn the switch ON/OFF remotely by sending commands to the device via MyQttHub. I can update the rules remotely, e.g. I needed to scale back the frequency of the reports from 300 sec to 900. The MyQttHub API calls to retrieve the data were added to my Windows-based C# client program where I collect data from my other humidity/temperature sensors at that location (will try to attach a graph sample later). There are two things I did not figure out yet (appreciate any ideas):
1. How to achieve a real-time control of the switch, like either set some rules at MyQttHub, or ask MyQttHub to forward the device reports to some other server what would maintain the rules and issue messages to the device based on some logic (time of day, other sensors, etc.). All I can do now is collect humidity measurements in bulk after the fact. I guess I could setup my own server to poll every 15 min (like 1 min behind the device reports) and read data from MyQttHub like I do now and then do what I need. However this is not very real-time and I would rather get events from MyQttHub than poll it.
2. A separate thing I need - can I program a timer rule on the device itself, in addition to the existing humidity-controlled switch rule? Like if I switch it ON remotely, it would turn OFF in 15 min. Or if I press a button on the switch, it would turn OFF by itself in 15 min. Currently the rule example provided in the switch documentation only works on humidity, so if I turn the fan ON manually, it keeps going forever. How many independent rules can a humidity switch like this handle?

**Reply by sun2sirius** [reply]
![SmartThings-Plot](https://github.com/arendst/Tasmota/assets/13130690/62542994-aa80-4739-a520-ad98751fef54)
I have two Cielo humidity sensors - red line in bedroom and blue line in living room. The black line is dehumidifier control, it turns on when bedroom humidity reaches 70%. Look how smooth is the green line I get over MQTT from the Tasmota sensor compared to Cielo sensors - very nice! You can also see how dehumidifier brings the humidity down and how the sensors react.

**Reply by barbudor** [reply]
For point 1) MQTT is realtime. So I don't get your point about getting the measurment after the fact.
If you setup for exemple a node-red or a C#/MQTT client connected to MyQTTHub, you would get data in real time and can react immediately.

For point 2) There a couple of options
- If you want the fan to be ON for 15 minutes whatever is the source of the POWER ON command, then `Pulsetime` is your friend.
 Once you have configured `Pulsetime` on an output, it will switchoff automatically after a delay. Issuing a new POWER ON before the end of the delay will retrigger the delay.
- If you want to specifically issue a "POWER ON FOR XXX MIN" command only on some cases, then command `TimedPower` is the way to go. It hasn't yet made it's way to doc despite being in Tasmota since 13.3.0.3
 `TimedPower<index> <milliseconds>[,ON|OFF|TOGGLE|BLINK]` executes `Power<index> [ON|OFF|TOGGLE|BLINK] ` and after <millisecond> executes `Power<index> [OFF|ON|TOGGLE|BLINK_OFF]`
  So for 15 minutes, the delay should be 900000

**Comment by sun2sirius** [comment]
For 1) - I think you may be talking about an MQTT-protocol client connected to MyQTTHub (MQTT broker). Then the client would probably open a connection to the broker and listen, getting the response immediately as soon as the broker receives it from the device and routes it to the client. I tried some of the MQTT clients, but was unable to establish a connection with MyQTTHub over MQTT, the only way I was able to use MyQTTHub is by connecting to it over their proprietary HTTP API to download stash. In other words, device MQTT messages are stashed at MyQTTHub, and my client downloads the stash over HTTP - that is why this method is not real-time.
I does look like my client is treated as just another device, so if the sensor can connect to MyQTTHub over MQTT, a Windows client should too using the same credentials and protocol. I just could not make it work for some reason. In any case, ideally I would need to bring the device feed from MyQTTHub into SmartThings somehow, because that is where all my homes, devices and rules are. Otherwise, I'll be building a separate and isolated subsystem...

**Reply by barbudor** [reply, marked as answer]
Yes, a MQTT client
You definitively should be able to create a "device" in MyQTTHub just like for your Tasmota, connect and then subscribe to topics.
Since I changed PC 18 months ago I didn't re-installed VisualStudio so I don't have any C# environment up and running
I see there are a couple of MQTT libs for C#:
- https://github.com/dotnet/MQTTnet
- https://www.hivemq.com/blog/implementing-mqtt-in-c-sharp/
- https://eclipse.dev/paho/index.php?page=clients/dotnet/index.php

I'm using Python with paho-mqtt library

Regarding the SmartThings side, I don't know then how you could push from MQTT to SmartThings
I don't know if this could help: https://community.smartthings.com/t/st-edge-tasmota-edge-for-sonoff-tuya-many-other-esp-wifi-http-devices-over-2000-tasmota-supported-devices/241493
